File: contest/c_niveau_audition/sol/c_niveau_audition.exs
# Solution by Thomas Lacroix

read_int_list = fn ->
  IO.read(:stdio, :line) |> String.trim() |> String.split(" ") |> Stream.map(&String.to_integer/1)
end

_nb_rows = IO.read(:stdio, :line)
row_capacities = read_int_list.() |> Stream.with_index()
_nb_spectators = IO.read(:stdio, :line)
hearing_levels = read_int_list.() |> Enum.frequencies()

case Enum.reduce_while(
       row_capacities,
       {hearing_levels, 0},
       fn {row_capacity, i}, {hearing_levels, remaining_seats} = _acc ->
         row = i + 1
         {nb_persons, new_levels} = Map.pop(hearing_levels, row, 0)

         if row_capacity + remaining_seats >= nb_persons do
           {:cont, {new_levels, remaining_seats + (row_capacity - nb_persons)}}
         else
           {:halt, :impossible}
         end
       end
     ) do
  :impossible ->
    IO.puts("IMPOSSIBLE")

  {hearing_levels, remaining_seats} ->
    remaining_people = Map.values(hearing_levels) |> Enum.sum()
    IO.puts(if remaining_seats >= remaining_people, do: "POSSIBLE", else: "IMPOSSIBLE")
end

File: contest/c_niveau_audition/sol/c_niveau_audition.rs
// Solution by Thomas Lacroix

use std::fmt::Debug;

fn read<T: std::str::FromStr>() -> T
where
    T::Err: Debug,
{
    let mut line = String::new();
    std::io::stdin().read_line(&mut line).unwrap();
    line.trim().parse().unwrap()
}

fn read_list<T: std::str::FromStr>() -> Vec<T>
where
    T::Err: Debug,
{
    let mut line = String::new();
    std::io::stdin().read_line(&mut line).unwrap();
    line.trim().split(" ").map(|s| s.parse().unwrap()).collect()
}

fn main() {
    let _ = read::<u32>();
    let mut rows = read_list::<u32>();
    let _ = read::<u32>();
    let mut hearing_levels = read_list::<u32>();
    hearing_levels.sort();

    let mut row = 0;
    let mut possible = true;
    'outer: for person in 0..hearing_levels.len() {
        while rows[row] == 0 {
            if row + 1 == rows.len() {
                possible = false;
                break 'outer;
            }
            row += 1
        }

        if row + 1 > hearing_levels[person] as usize {
            possible = false;
            break 'outer;
        }

        rows[row] -= 1;
    }

    println!("{}POSSIBLE", if possible { "" } else { "IM" });
}
